services/gundeck/test/unit/MockGundeck.hs

```diff
@@ -0,0 +1,676 @@
+    forM (zip uids cidss) $ \(uid, cids) -> (uid,) <$> do
+      forM cids $ \cid -> (cid,) <$> genClientInfo uid cid
+
+  validateMockEnv env & either error (const $ pure env)
+
+-- Try to shrink a 'MockEnv' by removing some users from '_meClientInfos'.
+shrinkMockEnv :: HasCallStack => MockEnv -> [MockEnv]
+shrinkMockEnv (MockEnv cis) =
+    MockEnv . Map.fromList <$>
+    filter (not . null) (shrinkList (const []) (Map.toList cis))
+
+validateMockEnv :: forall m. MonadError String m => MockEnv -> m ()
+validateMockEnv env = do
+  checkIdsInNativeAddresses
+  -- (if you want to vaidate anything else, here is the place!)
+  where
+    -- UserId and ClientId contained in Address must match the keys under which they are stored.
+    checkIdsInNativeAddresses :: m ()
+    checkIdsInNativeAddresses = do
+      forM_ (Map.toList $ env ^. meClientInfos) $ \(uid, cinfos) -> do
+        forM_ (Map.toList cinfos) $ \(cid, cinfo) -> do
+          forM_ (cinfo ^. ciNativeAddress) $ \(adr, _) -> do
+            unless (uid == adr ^. addrUser && cid == adr ^. addrClient) $ do
+              throwError (show (uid, cid, adr))
+
+genRecipients :: HasCallStack => Int -> MockEnv -> Gen [Recipient]
+genRecipients numrcp env = do
+  uids  <- take numrcp <$> shuffle (allUsers env)
+  genRecipient' env `mapM` uids
+
+genRecipient :: HasCallStack => MockEnv -> Gen Recipient
+genRecipient env = do
+  uid   <- QC.elements (allUsers env)
+  genRecipient' env uid
+
+genRecipient' :: HasCallStack => MockEnv -> UserId -> Gen Recipient
+genRecipient' env uid = do
+  route <- genRoute
+  cids  <- sublistOf (clientIdsOfUser env uid)
+  pure $ Recipient uid route cids
+
+-- REFACTOR: see 'Route' type about missing 'RouteNative'.
+genRoute :: HasCallStack => Gen Route
+genRoute = QC.elements [RouteAny, RouteDirect]
+
+genId :: Gen (Id a)
+genId = do
+  gen <- mkStdGen <$> arbitrary
+  pure . Id . fst $ random gen
+
+genClientId :: Gen ClientId
+genClientId = newClientId <$> arbitrary
+
+genProtoAddress :: HasCallStack => UserId -> ClientId -> Gen (Address "no-keys")
+genProtoAddress _addrUser _addrClient = do
+  _addrTransport :: Transport <- QC.elements [minBound..]
+  arnEpId :: Text <- arbitrary
+  let _addrApp = "AppName"
+      _addrToken = Token "tok"
+      _addrEndpoint = mkFakeAddrEndpoint (arnEpId, _addrTransport, _addrApp)
+      _addrConn = fakeConnId _addrClient
+  pure Address {..}
+
+genPushes :: MockEnv -> Gen [Push]
+genPushes = listOf . genPush
+
+genPush :: HasCallStack => MockEnv -> Gen Push
+genPush env = do
+  let alluids = allUsers env
+  sender <- QC.elements alluids
+  rcps :: Range 1 1024 (Set Recipient) <- do
+    numrcp <- choose (1, min 1024 (length alluids))
+    rcps   <- genRecipients numrcp env
+    unsafeRange . Set.fromList <$> dropSomeDevices `mapM` rcps
+  pload <- genPayload
+  inclorigin <- arbitrary
+  transient <- arbitrary
+  pure $ newPush sender rcps pload
+    -- TODO: & pushConnections .~ _
+    -- TODO: & pushOriginConnection .~ _
```

Comment:
I think at least this one should be done before it's merged, no idea about `pushConnections`.